Add translation and scaling for 3D image layers

jnk22 opened the pull request:
This PR adds translation and scaling to 3D layers:

- **Translation** is supported for all axes using numeric input fields with a step size of 1.0 and a range of -inf to inf.
- **Scaling** supports one input that scales on all axes with a step size of 0.1.

I have set `0.00001 = 1e-5` as minimum scaling factor both for the input transformation as well as when recovering the scaling factor from the active layer. This should prevent any division-by-zero errors and ensure that rotation matrices are always valid.

The transformation order is the default order that is represented by the affine transformation matrix and how that is interpreted by napari. _(I might want to add the capability to the order configurable in the future.)_

![image](https://github.com/user-attachments/assets/dfd352d8-ee3f-4405-8434-ef7872eb2f05)

Apart from the changes above I have renamed **Rotation[Model|Widget|Component]** to **Transformation[Model|Widget|Component]** to reflect the general transformation capabilities.

The `TransformationModel.reset()` resets the whole transformation in one go without emitting value changes in between.

---

I have tested the changes by running the sample script in `src/napari_manual_transforms/_widget.py` and doing changes interactively.

tlambert03 commented on src/napari_manual_transforms/__init__.py:
```diff
@@ -6,6 +6,6 @@
+from ._widget import TransformationWidget
 
-__all__ = ["RotationWidget"]
+__all__ = ["TransformationWidget"]
```
I don't imagine anyone was directly importing this widget... but just in case, we need to deprecate the name before removing it

```suggestion
__all__ = ["TransformationWidget"]

def __getattr__(name: str) -> object:
    if name == "RotationWidget":
        from warnings import warn
        warnings.warn(
            "'RotationWidget' has been renamed to 'TransformationWidget'.",
            DeprecationWarning,
            stacklevel=2,
        )
        return TransformationWidget
    raise AttributeError(f"module {__name__} has no attribute {name}")
```

jnk22 replied:
I have implemented your suggestion.

tlambert03 commented on src/napari_manual_transforms/_model.py:
```diff
@@ -165,6 +179,26 @@ def compact_axis_angle(self):
+    @translation.setter
+    def translation(self, t: Sequence[float] | np.ndarray):
+        """x,y,z"""
+        self._t = np.asarray(t)
```
it looks like a strict assumption is made in `self.transform` that `self.translation` has three items correct?  If so, then we need some sort of validation to be done here in the setter.

jnk22 replied:
Yes, that is correct. I now convert the input to an array and check its dimension/size afterwards.
If it does not have a length of 3, a ValueError will be raised.

Also removed the type hint for allowing np.ndarrays here as quaternion inputs are similar to translation inputs but also only expect inputs of type `Sequence`.

tlambert03 commented on src/napari_manual_transforms/_model.py:
```diff
@@ -165,6 +179,26 @@ def compact_axis_angle(self):
+    @scale.setter
+    def scale(self, s: float):
+        """scale"""
+        self._s = s
```
similar, here. need some simple validation

jnk22 replied:
I have implemented a simple user warning with automatically adjusting the value to a minimum value.
Would you rather raise an exception here instead?

tlambert03 commented on src/napari_manual_transforms/_model.py:
```diff
@@ -177,11 +211,19 @@ def origin(self, v) -> None:
     def transform(self) -> np.ndarray:
         M = np.eye(4)
-        M[:3, :3] = self.matrix
+        M[:3, :3] = self.matrix @ np.diag((max(self.scale, MINIMUM_SCALE),) * 3)
```
seems like this `max()` could go in the setter... along with a warning if someone tries to set scale to 0.

jnk22 replied:
This should be resolved as validation now happens during the scale-setter.

tlambert03 commented on src/napari_manual_transforms/_widget.py:
```diff
@@ -158,7 +159,27 @@ def _on_layer_event(self, event):
+
+                # Prevent division by zero by using at least the minimum
+                # scaling value that is allowed as scaling input.
+                self._model.matrix = affine_matrix[:3, :3] / max(scale, MINIMUM_SCALE)
```
this `max()` and import of `MINIMUM_SCALE` could be removed if you validate the setting of `_model.scale`

jnk22 replied:
Same here, resolved imo.

tlambert03 reviewed (commented):
thanks @jnk22!  some minor questions/thoughts

tlambert03 commented:
> I have tested the changes by running the sample script in src/napari_manual_transforms/_widget.py and doing changes interactively.

that's fine with me, since there really isn't much of a test suite in this repo yet anyway.  but, if you do have time, and an idea about what you might test, you could add a test here (optional)

jnk22 commented:
Hi, thanks for your great feedback!

Just wanted to say that I have not forgotten about this PR and will work on it as soon as I have the time for it (hopefully within the next ~2 weeks).

I will try to find a way to implement at least some tests for my changes and also update the screenshot for the README as well, which I had forgotten before.

jnk22 commented:
Updated the code and included some tests for the transformation model class (mainly regarding the translation/scale stuff that I have added).

When creating a new screenshot, I have also noticed that the input rotations that I re-created from the current screenshot do not result in the exact same rotation as the screenshot shows. I have checked against current main, same behavior there, but double-checking might be good here.

The **resample**-button that I have missed before only respects rotation for now -- disregarding any translation and scaling. Should this also be included within this PR? Otherwise I could add another PR when I have some more time for it.
